### PR TITLE
Simplify evaluation workflow by removing benchmarks build polling

### LINKED ISSUE #1265: Improve Cross-Repository Evaluation Workflow Orchestration - Second Attempt
Current State
Architecture Overview
The evaluation pipeline currently spans three repositories with orchestration split between GitHub Actions workflows:

software-agent-sdk (GitHub Actions)
├── 1. Validates inputs & resolves SDK commit SHA
├── 2. Dispatches benchmarks image build (workflow_dispatch to benchmarks repo)
├── 3. Polls for build completion
│   ├── 60-second polling interval
│   ├── Up to 600 attempts (10 hours max wait)
│   ├── Timestamp-based run identification
│   └── 40+ GitHub API calls per evaluation
└── 4. Dispatches evaluation workflow (workflow_dispatch to evaluation repo)
    └── evaluation repo creates Kubernetes jobs
        └── run_eval.sh orchestrates:
            - Clone repos & install dependencies
            - Run SWE-bench inference
            - Execute evaluation harness
            - Upload results to GCS
Pain Points
1. Complexity Split Across Multiple Systems

Orchestration logic split between YAML (software-agent-sdk), bash (evaluation), and two repositories
327-line GitHub Actions workflow with ~100 lines dedicated to polling logic
Difficult to understand the full evaluation flow
2. Inefficient Polling Mechanism

Makes 40+ GitHub API calls to track a single build
Timestamp-based run identification vulnerable to race conditions with concurrent triggers
3. Limited Error Handling

YAML workflows provide minimal debugging information
Hard to add sophisticated retry logic or error recovery
Stack traces and logs scattered across multiple workflow runs
4. Maintainability Challenges

YAML is difficult to test (no unit tests)
Complex conditional logic hard to reason about
Adding features (notifications, retry strategies, metrics) requires more YAML complexity
5. GitHub Actions Resource Usage

Workflow runs consume Actions minutes just to poll another workflow
Could hit rate limits during concurrent evaluations
Proposed Solution
Architecture Overview
Move orchestration logic into the evaluation repository's Kubernetes job, where it naturally belongs:

software-agent-sdk (GitHub Actions - SIMPLIFIED)
├── 1. Validates inputs & resolves SDK commit SHA
└── 2. Dispatches evaluation workflow with orchestration parameters
    
evaluation repo (GitHub Actions)
└── Creates Kubernetes orchestration job
    └── orchestrate_eval.py:
        ├── 1. Trigger benchmarks build via GitHub API
        ├── 2. Launch evaluation (existing run_eval.sh logic)
        └──3. Report results

### PR DESCRIPTION
**Pull request by simonrosenberg**
# Simplify Cross-Repository Evaluation Workflow Orchestration

Fixes #1265

## Summary

This PR simplifies the SDK evaluation workflow by moving orchestration complexity from GitHub Actions (YAML) into Python code in the evaluation repository where it naturally belongs.

### Before
- SDK workflow: 327-line YAML with complex polling logic (~100 lines)
- Made 40+ GitHub API calls per evaluation to track benchmarks build
- Timestamp-based run identification (vulnerable to race conditions)
- Orchestration split between YAML, bash, and multiple repositories

### After
- SDK workflow: Simplified to ~200 lines
- Removed all polling logic
- Single dispatch to evaluation repo with all parameters
- Evaluation repo handles orchestration in Python (testable, maintainable)

#### 1. Simplified Workflow (`run-eval.yml`)
**Removed:**
- 100+ lines of benchmarks build polling logic
- GitHub API rate limit concerns
- Complex timestamp-based run identification
- Wait/retry loops

**Streamlined to:**
1. Validate inputs & authorization
2. Resolve SDK commit SHA
3. Resolve model configurations from `models.json`
4. Dispatch evaluation workflow with all parameters

#### 2. Model Configuration Management
**Created `resolve_model_configs.py`:**
- Validates model IDs against allowed list in `models.json`
- Resolves full model configurations including:
  - LiteLLM model identifiers
  - Temperature and other model-specific parameters
- Returns properly formatted JSON for evaluation workflow
- Handles both direct model IDs and comma-separated lists

**Replaced model stubs with single source of truth:**
- Removed `.github/run-eval/llm_config_model_stubs.json`
- Now using `.github/run-eval/models.json` as the single authoritative source
- Simplifies maintenance: one file to update for all model configurations

**Benefits:**
- Single file to maintain for model configs
- No risk of inconsistency between stubs and full configs
- Easier to add new models
- Validation happens before evaluation starts

#### 3. Enhanced Input Validation
**Added authorization check:**
- Only authorized users (listed in `authorized-labelers.txt`) can trigger evaluations
- Applies to both `workflow_dispatch` and `pull_request_target` events
- Prevents unauthorized resource usage

**Model ID validation:**
- Validates comma-separated model IDs against `models.json`
- Provides clear error messages with available options
- Fails fast before expensive evaluation starts

#### 4. Testing Feature Branches
**New `eval_branch` input:**
- Allows testing evaluation workflow changes before merging
- Defaults to `main` for normal operations
- Enables safe development and validation of orchestration changes

#### 5. Trigger Reason Propagation
**New `reason` input:**
- Accepts natural language descriptions with spaces
- Automatically sanitized for Helm compatibility (spaces → hyphens)
- Passed to evaluation workflow for Slack notifications and logging
- Provides context for why evaluations were triggered

## Code Review Updates

After initial implementation, addressed code review feedback to improve reliability:

### 1. Fixed Workflow Run Detection (HIGH PRIORITY)
**Problem:** Timestamp-based detection was unreliable and vulnerable to race conditions

**Solution:** Implemented baseline run ID comparison using Check Suites API
- Get baseline run ID before triggering workflow
- After trigger, poll for new run where `check_suite.id > baseline_id`
- Deterministic and race-condition-proof
- No reliance on timestamps

### 2. Removed Sanitize-Inputs Job (MEDIUM PRIORITY)
**Problem:** Unnecessary GitHub Actions job that could be done in Python

**Solution:** Moved sanitization logic to `orchestrate_eval.py`
- Trigger reason now sanitized in Python before helm deployment
- Reduces workflow complexity
- More testable and maintainable

### 3. Added Comprehensive Tests
- Unit tests for input sanitization
- Tests for baseline run ID detection logic
- Edge case handling

### Full Workflow Testing

**Test 1: Valid Model ID**
- Run: https://github.com/OpenHands/software-agent-sdk/actions/runs/19703922484
- Result: ✅ SUCCESS
- Validated: Model ID validation, authorization, workflow dispatch

**Test 2: Evaluation Deployment**
- Run: https://github.com/OpenHands/evaluation/actions/runs/19703926515
- Result: ✅ SUCCESS
- Validated:
  - Kubernetes deployment successful
  - Trigger reason sanitization working
  - TRIGGER_REASON environment variable passed to pod
  - Full orchestration flow complete

**Test 3: Invalid Model ID (Validation)**
- Run: https://github.com/OpenHands/software-agent-sdk/actions/runs/19703857863
- Result: ✅ EXPECTED FAILURE - "Model ID 'OpenHands/claude-haiku-4-20251001' not found"
- Validated: Input validation working correctly

**Test 4: Code Review Fixes - End-to-End**
- SDK Run: https://github.com/OpenHands/software-agent-sdk/actions/runs/19708030468
- Evaluation Run: https://github.com/OpenHands/evaluation/actions/runs/19708035858
- Result: ✅ SUCCESS
- Validated:
  - Baseline run ID detection working correctly
  - Sanitization in Python working correctly
  - Helm deployment successful with sanitized trigger_reason
  - Complete orchestration flow from SDK → Evaluation → Kubernetes

### Trigger Reason Feature
**Input:** "Testing fix for helm deployment issue"
**Sanitized:** "Testing-fix-for-helm-deployment-issue"
**Result:** ✅ Successfully passed to Kubernetes pod via helm_args

### Reduced Complexity
- **SDK workflow**: 327 → ~200 lines (-40%)
- **Polling logic**: Removed entirely
- **API calls**: 40+ → 1 per evaluation

### Improved Maintainability
- Orchestration logic now in Python (testable)
- Single source of truth for model configs
- Clear separation of concerns
- Deterministic workflow run detection

### Better Error Handling
- Fast-fail validation before expensive operations
- Clear error messages
- Authorization checks

### Enhanced Features
- Support for testing feature branches
- Trigger reason tracking
- Model ID validation

## Testing

- [x] SDK workflow triggers successfully
- [x] Model ID validation works
- [x] Authorization checks function correctly
- [x] Evaluation workflow receives correct parameters
- [x] Kubernetes deployment succeeds
- [x] Trigger reason sanitization and propagation works
- [x] Invalid inputs are properly rejected
- [x] Baseline run ID detection works correctly
- [x] Python-based sanitization works correctly
- [x] End-to-end workflow with code review fixes succeeds

## Related PRs

- Evaluation repo: https://github.com/OpenHands/evaluation/pull/55

---

**Co-authored-by:** openhands <openhands@all-hands.dev>

---
**Agent Server images for this PR**

• **GHCR package:** https://github.com/OpenHands/agent-sdk/pkgs/container/agent-server

**Variants & Base Images**
| Variant | Architectures | Base Image | Docs / Tags |
|---|---|---|---|
| java | amd64, arm64 | `eclipse-temurin:17-jdk` | [Link](https://hub.docker.com/_/eclipse-temurin:17-jdk) |
| python | amd64, arm64 | `nikolaik/python-nodejs:python3.12-nodejs22` | [Link](https://hub.docker.com/_/nikolaik/python-nodejs:python3.12-nodejs22) |
| golang | amd64, arm64 | `golang:1.21-bookworm` | [Link](https://hub.docker.com/_/golang:1.21-bookworm) |

**Pull (multi-arch manifest)**
```bash
# Each variant is a multi-arch manifest supporting both amd64 and arm64
docker pull ghcr.io/openhands/agent-server:0f06ad7-python
```

**Run**
```bash
docker run -it --rm \
  -p 8000:8000 \
  --name agent-server-0f06ad7-python \
  ghcr.io/openhands/agent-server:0f06ad7-python
```

**All tags pushed for this build**
```
ghcr.io/openhands/agent-server:0f06ad7-golang-amd64
ghcr.io/openhands/agent-server:0f06ad7-golang_tag_1.21-bookworm-amd64
ghcr.io/openhands/agent-server:0f06ad7-golang-arm64
ghcr.io/openhands/agent-server:0f06ad7-golang_tag_1.21-bookworm-arm64
ghcr.io/openhands/agent-server:0f06ad7-java-amd64
ghcr.io/openhands/agent-server:0f06ad7-eclipse-temurin_tag_17-jdk-amd64
ghcr.io/openhands/agent-server:0f06ad7-java-arm64
ghcr.io/openhands/agent-server:0f06ad7-eclipse-temurin_tag_17-jdk-arm64
ghcr.io/openhands/agent-server:0f06ad7-python-amd64
ghcr.io/openhands/agent-server:0f06ad7-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-amd64
ghcr.io/openhands/agent-server:0f06ad7-python-arm64
ghcr.io/openhands/agent-server:0f06ad7-nikolaik_s_python-nodejs_tag_python3.12-nodejs22-arm64
ghcr.io/openhands/agent-server:0f06ad7-golang
ghcr.io/openhands/agent-server:0f06ad7-java
ghcr.io/openhands/agent-server:0f06ad7-python
```

**About Multi-Architecture Support**
- Each variant tag (e.g., `0f06ad7-python`) is a **multi-arch manifest** supporting both **amd64** and **arm64**
- Docker automatically pulls the correct architecture for your platform
- Individual architecture tags (e.g., `0f06ad7-python-amd64`) are also available if needed

### CODE REVIEW

File: .github/run-eval/allowed-model-stubs.json
Comment (enyst) [inline .github/run-eval/allowed-model-stubs.json:1]: 😅 happy to see it gone!

File: .github/workflows/run-eval.yml [hunk 1]
```diff
@@ -176,110 +182,24 @@ jobs:
...
-                      -H "Accept: application/vnd.github+json" \
-                      "https://api.github.com/repos/${REPO}/actions/workflows/${WORKFLOW_ID}/runs?per_page=5&branch=main&created=>=$DISPATCH_TIME")
-
-                    # Find the first run created at or after dispatch time
-                    if [ -z "$RUN_ID" ]; then
-                      RUN_ID=$(echo "$RUNS" | jq -r ".workflow_runs[] | select(.created_at >= \"$DISPATCH_TIME\") | .id" | head -1)
-                      if [ -n "$RUN_ID" ] && [ "$RUN_ID" != "null" ]; then
-                        echo "Found workflow run: $RUN_ID"
-                      else
-                        echo "Waiting for workflow run to appear (attempt $i/$MAX_ATTEMPTS)..."
```
Comment (enyst) [inline .github/workflows/run-eval.yml:244]: Out of curiosity, what happens with all this code building from benchmarks, how will it work now?

**Comment by enyst** [comment]
I think I asked this before, sorry, but I just checked yesterday's PR, and I didn't see any comment in response to `eval-1` label?

File: .github/run-eval/resolve_model_configs.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    resolved = []
+    for model_id in model_ids:
+        if model_id not in models_by_id:
+            error_exit(f"Model ID '{model_id}' not found")
+        resolved.append(models_by_id[model_id])
+    return resolved
+
+
+def main() -> None:
+    models_json_path = get_required_env("MODELS_JSON_PATH")
```
Comment (xingyaoww) [inline .github/run-eval/resolve_model_configs.py]: How about we write models.json as a dict explicitly in this file since it seems it is only used here?

File: .github/workflows/run-eval.yml [hunk 3]
```diff
@@ -73,32 +69,33 @@ jobs:
...
-              if: github.event_name == 'pull_request_target'
+            - name: Validate authorization
+              if: github.event_name == 'pull_request_target' || github.event_name == 'workflow_dispatch'
               run: |
-                  LABELER="${{ github.actor }}"
-                  if ! grep -Fx "$LABELER" .github/run-eval/authorized-labelers.txt >/dev/null; then
-                    echo "User $LABELER is not authorized to trigger eval." >&2
+                  ACTOR="${{ github.actor }}"
+                  if ! grep -Fx "$ACTOR" .github/run-eval/authorized-actors.txt >/dev/null; then
+                    echo "User $ACTOR is not authorized to trigger eval." >&2
```
Comment (xingyaoww) [inline .github/workflows/run-eval.yml]: Maybe we can just remove this authorization? Only maintainer/ at least people with triage access to the repo can tag a PR, so i think it should be fine even if we don't check for `github.action`

File: .github/workflows/run-eval.yml (hunk 1 above)
Comment (xingyaoww) [reply .github/workflows/run-eval.yml:244]: We moved it to the Kubernetes workflow in our internal `OpenHands/evaluation` repo 😓 We don't want to keep polling stuff in CI since CI compute is likely more expensive than K8S, and keep these centralized in one place is a bit easier to maintain in the long run anyway

File: .github/workflows/run-eval.yml (hunk 3 above)
Comment (enyst) [reply .github/workflows/run-eval.yml]: Just to note. This might have consequences the other way around though: we basically don't have new maintainers from the community for a long time. If we add an additional criterium, it might become a little harder even?

Comment (enyst) [reply .github/workflows/run-eval.yml]: This goes both ways. I don't know. How about we let it sink in and give it some thought for a while?

File: .github/workflows/run-eval.yml (hunk 1 above)
Comment (enyst) [reply .github/workflows/run-eval.yml:244]: That's fine if the results are transparent and reproducible IMHO

**Comment by enyst** [comment]
> **Evaluation Triggered**
> 
> * Trigger: Label 'run-eval-1' on PR [Simplify evaluation workflow by removing benchmarks build polling #1267](https://github.com/OpenHands/software-agent-sdk/pull/1267)
> * SDK: [fc143ed](https://github.com/OpenHands/software-agent-sdk/commit/fc143edc9abcda3c307d83ffe496cc28821ba15c)
> * Eval limit: 1
> * Models: claude-sonnet-4-5-20250929

It still didn’t post a comment. Shouldn’t we expect it to?

**Comment by simonrosenberg** [comment]
> > **Evaluation Triggered**
> > 
> > * Trigger: Label 'run-eval-1' on PR [Simplify evaluation workflow by removing benchmarks build polling #1267](https://github.com/OpenHands/software-agent-sdk/pull/1267)
> > * SDK: [fc143ed](https://github.com/OpenHands/software-agent-sdk/commit/fc143edc9abcda3c307d83ffe496cc28821ba15c)
> > * Eval limit: 1
> > * Models: claude-sonnet-4-5-20250929
> 
> It still didn’t post a comment. Shouldn’t we expect it to?

let me rerun it with the latest code changes and check if the flow works!

**Comment by simonrosenberg** [comment]
> > **Evaluation Triggered**
> > 
> > * Trigger: Label 'run-eval-1' on PR [Simplify evaluation workflow by removing benchmarks build polling #1267](https://github.com/OpenHands/software-agent-sdk/pull/1267)
> > * SDK: [fc143ed](https://github.com/OpenHands/software-agent-sdk/commit/fc143edc9abcda3c307d83ffe496cc28821ba15c)
> > * Eval limit: 1
> > * Models: claude-sonnet-4-5-20250929
> 
> It still didn’t post a comment. Shouldn’t we expect it to?

Because of `pull_request_target` it's the base branch workflow that runs, not the feature branch workflow

**Comment by simonrosenberg** [comment]
> > **Evaluation Triggered**
> > 
> > * Trigger: Label 'run-eval-1' on PR [Simplify evaluation workflow by removing benchmarks build polling #1267](https://github.com/OpenHands/software-agent-sdk/pull/1267)
> > * SDK: [fc143ed](https://github.com/OpenHands/software-agent-sdk/commit/fc143edc9abcda3c307d83ffe496cc28821ba15c)
> > * Eval limit: 1
> > * Models: claude-sonnet-4-5-20250929
> 
> It still didn’t post a comment. Shouldn’t we expect it to?

To debug the PR commenting logic, I will temporarily update the workflow_dispatch logic to accept a PR number parameter and pretend that it was triggered from a label. We'll see if the eval commenting works.

**Comment by simonrosenberg** [comment]
> ## 🎉 Evaluation Job Completed
> **Evaluation Name:** `sdk-main-19764754304-claude-son` **Model:** `litellm_proxy/claude-sonnet-4-5-20250929` **Dataset:** `princeton-nlp/SWE-bench_Verified` (`test`) **Commit:** `45afbc3046d90b78918ec12ba14229c24887eb5e` **Timestamp:** 2025-11-28 13:26:16 UTC **Reason:** Testing-temporary-branch-workflow-matching-implementation
> 
> ### Results Summary
> * **Total instances:** 500
> * **Submitted instances:** 1
> * **Resolved instances:** 1
> * **Unresolved instances:** 0
> * **Empty patch instances:** 0
> * **Error instances:** 0
> * **Success rate:** 1/1 (100.0%)
> 
> View Metadata | View Results | Download Full Results

@enyst  FYI

**Comment by enyst** [comment]
> > ## 🎉 Evaluation Job Completed
> > **Evaluation Name:** `sdk-main-19764754304-claude-son` **Model:** `litellm_proxy/claude-sonnet-4-5-20250929` **Dataset:** `princeton-nlp/SWE-bench_Verified` (`test`) **Commit:** `45afbc3046d90b78918ec12ba14229c24887eb5e` **Timestamp:** 2025-11-28 13:26:16 UTC **Reason:** Testing-temporary-branch-workflow-matching-implementation
> > ### Results Summary
> > 
> > * **Total instances:** 500
> > * **Submitted instances:** 1
> > * **Resolved instances:** 1
> > * **Unresolved instances:** 0
> > * **Empty patch instances:** 0
> > * **Error instances:** 0
> > * **Success rate:** 1/1 (100.0%)
> > 
> > View Metadata | View Results | Download Full Results
> 
> @enyst FYI

Thank you, this is starting to work! It doesn't have any links to "Download Full Results" and the rest, though? They look like they would be links, but they're not.

We need links to look into results, IMHO, numbers alone don't tell much. Specially for debugging bad results or simply understanding them.

_Edited to add:_
For example, [like we had before](https://github.com/OpenHands/OpenHands/pull/9140#issuecomment-2974227429) at least?

**Comment by simonrosenberg** [comment]
> ## 🎉 Evaluation Job Completed
> **Evaluation Name:** `sdk-main-19769532756-claude-son` **Model:** `litellm_proxy/claude-sonnet-4-5-20250929` **Dataset:** `princeton-nlp/SWE-bench_Verified` (`test`) **Commit:** `e5bbafee3c19cde57a57bfe41f9b9b256043e9e6` **Timestamp:** 2025-11-28 17:02:06 UTC **Reason:** Test-clickable-GCS-links-in-PR-comments
> 
> ### Results Summary
> * **Total instances:** 500
> * **Submitted instances:** 1
> * **Resolved instances:** 1
> * **Unresolved instances:** 0
> * **Empty patch instances:** 0
> * **Error instances:** 0
> * **Success rate:** 1/1 (100.0%)
> 
> [View Metadata](https://console.cloud.google.com/storage/browser/_details/openhands-evaluation-results/artifacts/sdk-main-19769532756-claude-son/metadata.json) | [View Results](https://console.cloud.google.com/storage/browser/_details/openhands-evaluation-results/artifacts/sdk-main-19769532756-claude-son/results_summary.json) | [Download Full Results](https://console.cloud.google.com/storage/browser/_details/openhands-evaluation-results/sdk-main-19769532756-claude-son_litellm_proxy-claude-sonnet-4-5-20250929_25-11-28-17-02.tar.gz)

@enyst

**Comment by enyst** [comment]
Unfortunately, those links are not public, not accessible outside AH I assume. 😢 
@simonrosenberg 

I believe that's why Mamoodi had made in the past a full log available via GitHub actions. Not sure it was a great way, but it was a way.

**Comment by enyst** [comment]
Just to note, it still doesn't like me. 😢 

> [View Metadata](https://storage.googleapis.com/openhands-evaluation-results/artifacts/sdk-main-19832354049-claude-son/metadata.json) | [View Results](https://storage.googleapis.com/openhands-evaluation-results/artifacts/sdk-main-19832354049-claude-son/results_summary.json) | [Download Full Results](https://storage.googleapis.com/openhands-evaluation-results/sdk-main-19832354049-claude-son_litellm_proxy-claude-sonnet-4-5-20250929_25-12-01-18-10.tar.gz)

<Error>
<Code>AccessDenied</Code>
<Message>Access denied.</Message>
<Details>Anonymous caller does not have storage.objects.get access to the Google Cloud Storage object. Permission 'storage.objects.get' denied on resource (or it may not exist).</Details>
</Error>

**Comment by simonrosenberg** [comment]
> Just to note, it still doesn't like me. 😢
> 
> > [View Metadata](https://storage.googleapis.com/openhands-evaluation-results/artifacts/sdk-main-19832354049-claude-son/metadata.json) | [View Results](https://storage.googleapis.com/openhands-evaluation-results/artifacts/sdk-main-19832354049-claude-son/results_summary.json) | [Download Full Results](https://storage.googleapis.com/openhands-evaluation-results/sdk-main-19832354049-claude-son_litellm_proxy-claude-sonnet-4-5-20250929_25-12-01-18-10.tar.gz)
> 
> `AccessDenied` Access denied.

It's work in progress!

**Comment by simonrosenberg** [comment]
@enyst  btw the URL fixing is done ins a complete other repo, if you want to review the current code t would be awesome :)

File: .github/workflows/run-eval.yml [hunk 4]
```diff
@@ -140,146 +143,56 @@ jobs:
...
               env:
-                  SDK_SHA: ${{ steps.get-sha.outputs.sdk_sha }}
+                  SDK_SHA: ${{ steps.params.outputs.sdk_sha }}
                   EVAL_LIMIT: ${{ steps.params.outputs.eval_limit }}
-                  MODELS: ${{ steps.params.outputs.models }}
+                  MODELS_JSON: ${{ steps.resolve-models.outputs.models_json }}
                   EVAL_REPO: ${{ env.EVAL_REPO }}
                   EVAL_WORKFLOW: ${{ env.EVAL_WORKFLOW }}
+                  EVAL_BRANCH: ${{ github.event.inputs.eval_branch || 'main' }}
+                  BENCHMARKS_BRANCH: ${{ github.event.inputs.benchmarks_branch || 'main' }}
```
Comment (enyst) [inline .github/workflows/run-eval.yml:182]: Out of curiosity, is a particular `benchmark branch` still in use?

Comment (simonrosenberg) [reply .github/workflows/run-eval.yml:182]: No it's just `main` right now. But since there's lots of activity in that repo I think it's nice to leave a simple way of doing e2e testing

File: .github/workflows/run-eval.yml [hunk 5]
```diff
@@ -69,36 +72,34 @@ jobs:
...
+                  # Extract all model IDs from resolve_model_config.py
+                  ALLOWED_MODEL_IDS=$(python3 << 'EOF'
+                  import sys
+                  sys.path.insert(0, '.github/run-eval')
+                  from resolve_model_config import MODELS
+                  import json
+                  print(json.dumps(list(MODELS.keys())))
+                  EOF
+                  )
+                  DEFAULT_MODEL=$(echo "$ALLOWED_MODEL_IDS" | jq -r '.[0]')
```
Comment (enyst) [inline .github/workflows/run-eval.yml:90]: If I understand correctly, [my little trick](https://github.com/OpenHands/software-agent-sdk/pull/1166/commits/53f03017ac151725595a3a8042f79262e490a4bf#diff-28a1d8c11bba860a0e51a0b7bdd1d24c1e7a1215d9570fcc4a75da5b6d92df4fR7) will work now? 😅

Comment (simonrosenberg) [reply .github/workflows/run-eval.yml:90]: yes! no more hidden safeguards protecting us from you :)

**Comment by simonrosenberg** [comment]
> At a cursory look, it seems to me like we could maybe simplify some code. Let's take it in though, and have some fun!
> 
> If it works now, we can always adjust things as they come.

thank you!!
